### PR TITLE
Run Task related refactoring

### PR DESCRIPTION
**Pull request by RomanNikitenko**
#### What it does
**1. Change the order of search task configuration at running**
At the moment at running a task the order of search is:
- search a config among `provided`/`detected` tasks
- search a config among `configured` tasks
- search a config among `customized` tasks

It works and sure it founds the corresponding config for running.
The problem is: it does extra requests when it tries to find a `configured` task configuration.

Let's say user runs some `configured` task:
- task service [tries to find](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-service.ts#L469) a config among `provided` tasks first
- it tries [to get `cached` provided task](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/provided-task-configurations.ts#L60) 
- sure, there is no `configured` task among `provided` tasks, so it tries to [get all provided tasks](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/provided-task-configurations.ts#L64)
- it gets all providers, [fetches and caches](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/provided-task-configurations.ts#L41-L53) all `provided` tasks
- it doesn't find the corresponding config, so next it tries to [get a config among `configured` tasks](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-service.ts#L471)

I think we can change the order and do search among `configured` tasks first. It will allow to avoid extra requests: it just check [if a task exists in map](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-configurations.ts#L184) - we don't read `tasks.json` file at this step, so it can not slow down a search.
So:   
- if we are looking for a `configured` task, it finds the corresponding config and do not search among `provided` tasks
- if we are looking for a `provided` task, it  just check [if a task exists in map](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-configurations.ts#L184), then it tries to find among `provided` tasks 

**2. Separate running a compound task** 
At the moment for any task config we do [extra logic related to running a compound task](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-service.ts#L509-L524).

But on the next step we need it [if a task is compound](https://github.com/eclipse-theia/theia/blob/4c7275761eba622edee6df2e44c9e9e25f9e4f96/packages/task/src/browser/task-service.ts#L532) only.

My propose is: separate this logic in another method and execute it if it's really required. 

#### How to test
It's just refactoring, so we should avoid a regression.
1. Try to run any `configured` task
2. Try to run any `detected` task
3. Try to run a `compound` task

For `compound` tasks testing I used in `tasks.json` file configurations like:
```
"tasks": [
    {
      "type": "shell",
      "label": "1111",
      "command": "sleep 2 && echo 111111",
      "dependsOn": "2222"
    },
    {
      "type": "shell",
      "label": "2222",
      "command": "sleep 2 && echo 222222",
      "dependsOn": "3333"
    },
    {
      "type": "shell",
      "label": "3333",
      "command": "sleep 2 && echo 33333"
    }
  ]
```

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by RomanNikitenko** [comment]
@elaihau 
could you take a look, is it important to keep the current order of search or we can change it
thanks in advance!

File: packages/task/src/browser/task-service.ts [hunk 1]
```diff
@@ -516,9 +528,7 @@ export class TaskService implements TaskConfigurationClient {
             this.messageService.error(error.message);
             return undefined;
         }
-        return this.runTasksGraph(task, tasks, {
-            customization: { ...customizationObject, ...{ problemMatcher: resolvedMatchers } }
-        }).catch(error => {
+        return this.runTasksGraph(task, tasks, option).catch(error => {
             console.log(error.message);
```
Comment (akosyakov) [inline packages/task/src/browser/task-service.ts:532]: I wonder why do we not log full stack trace here and why we don't use `console.error`?  The same above at `this.logger.error(error.message)`.

File: packages/task/src/browser/task-service.ts [hunk 2]
```diff
@@ -506,8 +506,20 @@ export class TaskService implements TaskConfigurationClient {
...
-        const tasks = await this.getWorkspaceTasks(task._scope);
         const resolvedMatchers = await this.resolveProblemMatchers(task, customizationObject);
+        const runTaskOption: RunTaskOption = {
+            customization: { ...customizationObject, ...{ problemMatcher: resolvedMatchers } }
+        };
+
+        if (task.dependsOn) {
+            return this.runCompoundTask(task, runTaskOption);
+        } else {
+            return this.runTask(task, runTaskOption);
```
Comment (akosyakov) [inline packages/task/src/browser/task-service.ts:517]: Does it change error handling semantic here? It seems before the method will return `undefined` but now it throws? or `runTask` never throws?

**Comment by akosyakov** [comment]
I think changes looks good. Not sure about the motivation. Is it performance optimisation? Usually we test them by looking at numbers before and after to understand whether refactoring is worth it.

**Comment by RomanNikitenko** [comment]
> Not sure about the motivation. Is it performance optimisation?

yes, the PR allows to:
- avoid execution `compound` task related logic for ordinary tasks. For example, why we first get all tasks for a workspace and on the next step check if it's really required.
- skip extra step at running `configured` task (get all providers, fetch and cache all provided tasks)

File: packages/task/src/browser/task-service.ts (hunk 1 above)
Comment (RomanNikitenko) [reply packages/task/src/browser/task-service.ts]: Fixed, please take a look https://github.com/eclipse-theia/theia/pull/7696/commits/a5367f9314321de98ccf3894b4aaea4d2652d91e

File: packages/task/src/browser/task-service.ts (hunk 2 above)
Comment (RomanNikitenko) [reply packages/task/src/browser/task-service.ts]: it catches errors at running a configuration [here](https://github.com/eclipse-theia/theia/blob/e31a23d6566ef4d2f92fb53f5280783b812e7de0/packages/task/src/browser/task-service.ts#L929-L932), but, taking into account that we have chaining of call at running task, I added the corresponding block: https://github.com/eclipse-theia/theia/pull/7696/commits/a5367f9314321de98ccf3894b4aaea4d2652d91e  
thank you!

**Comment by RomanNikitenko** [comment]
> Not sure about the motivation. Is it performance optimisation? Usually we test them by looking at numbers before and after to understand whether refactoring is worth it.

@akosyakov 
I did some measure using `console.time()-console.timeEnd()` for simple (without `dependsOn`) `configured` `shell` and `detected` `tsc` tasks. I used repo from `How to test` section of the PR https://github.com/eclipse-theia/theia/pull/7676 to have a lot of tasks.

I had 3 points for `master` branch:
- [get config](https://github.com/eclipse-theia/theia/blob/7498bd1744fb00e583a8712220bfa615e54c669e/packages/task/src/browser/task-service.ts#L469-L479)
- [get workspace tasks](https://github.com/eclipse-theia/theia/blob/master/packages/task/src/browser/task-service.ts#L509)
- run: between [starting point](https://github.com/eclipse-theia/theia/blob/7498bd1744fb00e583a8712220bfa615e54c669e/packages/task/src/browser/task-service.ts#L468) and [this point](https://github.com/eclipse-theia/theia/blob/7498bd1744fb00e583a8712220bfa615e54c669e/packages/task/src/browser/task-service.ts#L697)

Please see details:
measure_way

I had only 2 points for my branch because `get workspace tasks` part is not involved in running `simple` (without `dependsOn`) task process after applying my changes.

**For `master` branch**
Detected task:
master_detected2

Configured task:
master_configured

**For my branch**
Detected task:
my_changes_detected

Configured task:
my_changes_configured

The difference is:
Detected task, master
master_detected
Detected task, my branch
my_changes_detected

Configured, master
master_configured
Configured, my branch
my_changes_configured

**Comment by RomanNikitenko** [comment]
Results, gifs

**For `master` branch**
Detected task:
master_detected2

Configured task:
master_configured

**For my branch**
Detected task:
my_changes_detected

Configured task:
my_changes_configured

**Comment by RomanNikitenko** [comment]
**The difference is:**

Detected task, master
master_detected
Detected task, my branch
my_changes_detected

Configured, master
master_configured
Configured, my branch
my_changes_configured

**Comment by RomanNikitenko** [comment]
So, you can see that:

For a `configured` task `get config` step reduced from ~300 ms to 0.035 ms because my changes allows to skip extra steps at running `configured` task (get all providers, fetch and cache all provided tasks) and just take the corresponding config from map.

For a `detected` task `get config` step we have the same time ~ 60 ms  - so after my changes we don't have some regression for getting a configuration of `detected` tasks 

For both `configured` and `detected` simple task time between [starting point](https://github.com/eclipse-theia/theia/blob/7498bd1744fb00e583a8712220bfa615e54c669e/packages/task/src/browser/task-service.ts#L468) and [this point](https://github.com/eclipse-theia/theia/blob/7498bd1744fb00e583a8712220bfa615e54c669e/packages/task/src/browser/task-service.ts#L697) was reduced from 700-1000 ms to 50-60 ms because my changes allows to avoid execution `compound` task related logic at running a `simple` task

According to the screenshots above you can see that we have huge time for getting workspace tasks - about 600-800 ms. So we should consider if we can reduce this time and improve our code.

p.s. I think that measurement depends on number of `detected` and `configured` tasks in a workspace, but it should reflect our problems in general.

@akosyakov @elaihau 
wdyt?
is it good way to detect problem places?
should we have an issue to investigate [get workspace tasks](https://github.com/eclipse-theia/theia/blob/master/packages/task/src/browser/task-service.ts#L509) delay as possible place for improvement?
please let me know if you can see something wrong in my measurement

**Comment by RomanNikitenko** [comment]
I need to have these changes merged to continue investigate and detect delay at running tasks, so I'm merging this PR.

But we can continue to discuss here the result of the measurement.

**Comment by tsmaeder** [comment]
@RomanNikitenko if I understand the reasoning correctly, the idea is that if we have a detected task, we will look for a customization for that task anyway, so it doesn't matter whether we find the customized task or the contributed task...it will result in the same task being executed, right?

**Comment by RomanNikitenko** [comment]
@tsmaeder 

There were two improvements:
- separation `compound` task related logic
- prevention extra steps at running a configured task

First one has influence on all `simple` tasks: `configured`, `detected` and `customized`.
Second one reduces search time for a `configured` task because:
> skip extra steps at running configured task (get all providers, fetch and cache all provided tasks)

Regarding
> the idea is that if we have a detected task, we will look for a customization for that task anyway, so it doesn't matter whether we find the customized task or the contributed task...it will result in the same task being executed, right?

Sorry, I don't have a ready answer...

**Pull request merged**